### PR TITLE
ARROW-15848: [Gandiva][C++] Fix function istrue and is not true

### PR DESCRIPTION
**Pull request by Johnnathanalmeida**
Modifying the behavior of the function to consider false when it comes to null.

More detail:
The functions "istrue" and "isfalse" were replace from null_if_null to null_never in gandiva.

### CODE REVIEW

File: cpp/src/gandiva/function_registry_common.h [hunk 1]
```diff
@@ -285,4 +285,13 @@ typedef std::unordered_map<const FunctionSignature*, const NativeFunction*, KeyH
 #define NUMERIC_BOOL_DATE_VAR_LEN_TYPES(INNER, NAME, ALIASES) \
   NUMERIC_BOOL_DATE_TYPES(INNER, NAME, ALIASES), VAR_LEN_TYPES(INNER, NAME, ALIASES)
 
+// Is True and Is Not True functions with validity that :
+// - NULL handling is of type NULL_NEVER
+//
+// The pre-compiled fn name includes the base name & input type name. istrue_boolean /
+// isfalse_boolean
+#define ISTRUE_ISFALSE_NULL_NEVER(NAME, ALIASES, IN_TYPE, OUT_TYPE)                  \
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/function_registry_common.h:293]: You can follow the example of `UNARY_SAFE_NULL_IF_NULL` and create a more generic macro, like `UNARY_SAFE_NULL_NEVER`, if it does not exist yet

File: cpp/src/gandiva/precompiled/arithmetic_ops.cc [hunk 2]
```diff
@@ -234,9 +234,13 @@ NUMERIC_TYPES(VALIDITY_OP, isnumeric, +)
 
 #undef VALIDITY_OP
 
-#define IS_TRUE_OR_FALSE_BOOL(NAME, TYPE, OP) \
-  FORCE_INLINE                                \
-  gdv_##TYPE NAME##_boolean(gdv_##TYPE in) { return OP in; }
+#define IS_TRUE_OR_FALSE_BOOL(NAME, TYPE, OP)                      \
+  gdv_##TYPE NAME##_boolean(gdv_##TYPE in, gdv_boolean is_valid) { \
+    if (!is_valid) {                                               \
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/precompiled/arithmetic_ops.cc:239]: I think you can just return:
```
return !is_valid && OP in;
```

And remove the if clause

**Review by anthonylouisbsb — CHANGES_REQUESTED** [review]

**Review by anthonylouisbsb — APPROVED** [review]

File: cpp/src/gandiva/precompiled/arithmetic_ops.cc [hunk 3]
```diff
@@ -234,20 +234,42 @@ NUMERIC_TYPES(VALIDITY_OP, isnumeric, +)
...
+#define IS_TRUE_OR_FALSE_BOOL(NAME, TYPE, OP)                      \
+  FORCE_INLINE                                                     \
+  gdv_##TYPE NAME##_boolean(gdv_##TYPE in, gdv_boolean is_valid) { \
+    return is_valid && OP in;                                      \
+  }
+
+#define IS_NOT_TRUE_OR_IS_NOT_FALSE_BOOL(NAME, TYPE, OP)           \
+  FORCE_INLINE                                                     \
+  gdv_##TYPE NAME##_boolean(gdv_##TYPE in, gdv_boolean is_valid) { \
+    return !is_valid ? true : OP in;                               \
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/precompiled/arithmetic_ops.cc:246]: I think you can rewrite it to `!is_valid || OP in;`

File: cpp/src/gandiva/precompiled/arithmetic_ops.cc [hunk 4]
```diff
@@ -234,20 +234,42 @@ NUMERIC_TYPES(VALIDITY_OP, isnumeric, +)
...
 IS_TRUE_OR_FALSE_BOOL(istrue, boolean, +)
 IS_TRUE_OR_FALSE_BOOL(isfalse, boolean, !)
 
-#define IS_TRUE_OR_FALSE_NUMERIC(NAME, TYPE, OP) \
-  FORCE_INLINE                                   \
-  gdv_boolean NAME##_##TYPE(gdv_##TYPE in) { return OP(in != 0 ? true : false); }
+#define IS_TRUE_OR_FALSE_NUMERIC(NAME, TYPE, OP)                   \
+  FORCE_INLINE                                                     \
+  gdv_boolean NAME##_##TYPE(gdv_##TYPE in, gdv_boolean is_valid) { \
+    return is_valid && OP(in != 0 ? true : false);                 \
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/precompiled/arithmetic_ops.cc:258]: I think you can just keep it as `is_valid && OP(in != 0)`

File: cpp/src/gandiva/precompiled/arithmetic_ops.cc [hunk 5]
```diff
@@ -234,20 +234,42 @@ NUMERIC_TYPES(VALIDITY_OP, isnumeric, +)
...
+  }
 
 NUMERIC_TYPES(IS_TRUE_OR_FALSE_NUMERIC, istrue, +)
 NUMERIC_TYPES(IS_TRUE_OR_FALSE_NUMERIC, isfalse, !)
 
+#define IS_NOT_TRUE_OR_IS_NOT_FALSE_NUMERIC(NAME, TYPE, OP)        \
+  FORCE_INLINE                                                     \
+  gdv_boolean NAME##_##TYPE(gdv_##TYPE in, gdv_boolean is_valid) { \
+    return !is_valid ? true : (OP(in != 0 ? true : false));        \
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/precompiled/arithmetic_ops.cc:267]: I think you can keep it as `!is_valid || OP(in != 0)`

**Comment by Johnnathanalmeida** [comment]
@github-actions crossbow submit java-jars

**Comment by anthonylouisbsb** [comment]
@github-actions crossbow submit java-jars

**Review by vvellanki — APPROVED** [review]

**Pull request closed without merging**